yaml/shared/src/main/scala/org/virtuslab/yaml/internal/load/parse/ParserImpl.scala

```diff
@@ -49,202 +51,203 @@ private enum Production:
  * ParseValue            ::= <ParseNode> <ParseOptKey>
  * ParseSequenceEntryOpt ::= epsilon | <ParseNode> ParseSequenceEntryOpt
 */
-object ParserImpl extends Parser:
+class ParserImpl(in: Tokenizer) extends Parser:
   import Production.*
 
-  override def getEvents(in: Tokenizer): Either[YamlError, List[Event]] =
+  private val productions = mutable.ArrayDeque(ParseStreamStart)
+
+  override def getEvents(): Either[YamlError, List[Event]] =
     @tailrec
     def loop(
-        acc: List[Event],
-        productionStack: List[Production]
+        acc: List[Event]
     ): Either[YamlError, List[Event]] = {
-      ParserImpl.getNextEvent(in, productionStack) match
-        case Right((event, next)) =>
-          if event != Event.StreamEnd then loop(acc :+ event, next)
-          else Right(acc :+ event)
+      getNextEvent() match
+        case Right(event) =>
+          if event != Event.StreamEnd then loop(event :: acc)
+          else Right((event :: acc).reverse)
         case Left(err) => Left(err)
     }
-    loop(Nil, List(ParseStreamStart))
+    loop(Nil)
 
-  type EventResult = Either[YamlError, (Event, List[Production])]
+  override def getNextEvent(): Either[YamlError, Event] =
```

Comment:
In any case, after the StreamEvent event, the parser will not return another event? I think we should use Option.

Reply:
I changed this behavior. Now, it's returning right with `Event.StreamEnd` value if there are no more events